### ISSUE TITLE
proper Unicode support for FilePath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7805]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#7805|
|Type|enhancement|
|Created|2015-03-08 12:16:21Z|
|Branch|https://github.com/twisted/twisted/tree/unicode-filepath-7805|

Currently, `twisted.python.filepath.FilePath` and `twisted.python.filepath.IFilePath` are informally defined as "strings".  This has been increasingly strictly interpreted to mean "bytes" as we have ported various things to Python 3, which is an incorrect interpretation of what a path is, and has limited functionality in some important ways.

Path names are text.  Humans must type them in on keyboards at command lines in various places, view them in output on terminals and in GUIs, and otherwise interact with them as text.

However, the kernel interfaces on UNIX which deal with path names all operate in terms of bytes.  This creates a long and unfortunate legacy of buggy nonsense that we all have to deal with.

However, only on Linux (and FreeBSD) does this create ambiguity.  On Windows, the kernel interfaces are in fact in terms of Unicode, and on OS X, the default filename path canonicalization rules give you usable UTF-8 99% of the time.

Even on Linux, there are well-defined rules for how you are supposed to treat path names when dealing with them as text.  You respect the locale, and encode things according to that locale.  This can fail, so you need to be able to fall back to bytes, and ideally fall back to bytes one segment at a time, so that file _names_ can be displayed even in a scenario where full _paths_ cannot be.

So: `IFilePath` implementations should support Unicode.

The good news is that it sort of does sometimes, by accident.  The bad news is that it does less and less as time goes on and we port more things to py3 and implicit coercion stops working.

Here's how that should work.

To maintain compatibility, we need to maintain the unfortunate `path` attribute, which is the trickiest part of the whole mess.  However, it's not as bad as it might seem.

If you construct a `FilePath` with text, you should get a text-mode `FilePath`.  Much like the way that the stdlib `os` and `os.path` modules support either text or bytes, and return the same type you gave them as input, so should `FilePath`.

You should be able to pass bytes or text to the methods which take an input: `child` and `sibling`, regardless of the mode of the object in question.  If you pass text to `child`, you should get a text-mode `FilePath` back, regardless of whether the object you're calling it on is in text or bytes mode.

`FilePath` should then also have an `asTextPath` and an `asBytesPath` method.  Given that `sys.getfilesystemencoding` tells us the "right" encoding to use here, we don't need to take an argument there.

The `path` attribute should have a setter that mutates the mode of an existing `FilePath` instance, since it was previously settable, but this should probably be deprecated.

There should be different underlying implementations of the type for different platforms, since different platforms work differently.

 - _On Windows_, all `FilePath` objects, whether they're in text mode or not, should always maintain an internal representation in unicode, since that's how we always ought to be talking to the kernel.  When converting the output of a method for a bytes-mode path, it should be encoded as UTF-8, to preserve as much information as possible.  As far as I can tell, MBCS encoding is simply lossy and there's no particular benefit to use it, unless you need compatibility with some _really_ old legacy windows software (in which case everything is almost certainly ASCII anyway, so UTF-8 will be good enough).

 - _On UNIX_, all `FilePath` objects, whether they're in bytes mode or not, should always maintain an internal representation in bytes, and decode on output.  When encoding a text-mode path for user consumption, we can use the `surrogateescape` error handler on python 3 to preserve information as much as possible.

This strategy will not support all the kinds of wacky shenanigans that you can get up to on a UNIX filesystem (for a simple example, a UTF-8-encoded root filesystem with a Shift-JIS removable volume mounted on a path within it that has non-ASCII characters in the name).  However, this strategy _is_ forward-compatible with such shenanigans: initially, a program that cares about correctness in that scenario can just be careful to use bytes `FilePath` instances, and in the future converting those into text `FilePath` instances can do crazy stuff like read `mtab` to determine encodings and reconstitute unicode path names.

Methods that return portions of a file name (`basename`, or the not-in-`IFilePath` `dirname` and `splitext`) can defer decoding to just the split-up portion.  On Windows we can split on `u'\\'` and on UNIX we can split on `b'/'` before attempting to decode, so errors in the other parts of the path don't need to affect it.

This should affect some other issues:

    - [#5203](https://github.com/twisted/twisted/issues/5203) of which this may or may not be a duplicate
    - [#2366](https://github.com/twisted/twisted/issues/2366) which may or may not be resolvable by changing the way some Windows paths are computed
    - [#10489](https://github.com/twisted/twisted/issues/10489) which this might just resolve as a matter of course
    - [#7166](https://github.com/twisted/twisted/issues/7166) which goes against this ticket and ask for FilePath to reject Unicode

Attachments:

* [unicode-value-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/aee/aee387a1df2f170fd74eb6149f966c46bba16619/392b5b2b094a522fdf71d7f0e7fefbe5a604cc94.patch) (9810 bytes) - added by adiroiban on 2015-03-20 09:47:24Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7805 7805
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_unicode_filepath_7805 branches/unicode-filepath-7805
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1425816981757585 1425816981757585
changetime__1427189936910494 1427189936910494
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_1"></a>|@hynek commented|
|-|-|

I feel like I should add http://lucumr.pocoo.org/2014/5/12/everything-about-unicode/ to the discussion which seems relevant to the plans and may help discussing the various issues.

**Comment by twisted-trac** [comment]
|<img alt="eddygeek's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|eddygeek commented|
|-|-|

From an outsider: It seems 3 of the 4 tickets remaining open in  https://twistedmatrix.com/trac/wiki/Plan/Python3 link to this ticket or [#5203](https://github.com/twisted/twisted/issues/5203) ... so, quite a blocker ;-)

As to what are valid paths, UTF-8 will work with Unix but Windows uses UCS-2, with subtle differences... Racket, Scheme 48, and Mozilla Rust languages stdlib chose WTF-8 encoding to deal with this: see [this rfc](https://github.com/rust-lang/rfcs/blob/master/text/0474-path-reform.md#representation-choices-unicode-and-normalization) for motivations and [here](http://simonsapin.github.io/wtf-8/) for the actual encoding.

It seems to me unicode strings are close enough, and would allow continuing with Python3 porting.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Yeah, in Twisted, "all roads lead to FilePath". :(

The current path attribute, for backwards compat, can be a property that just calls asBytesPath(), and throws up if you try and set it (since that should NOT be part of the interface, a FilePath seems like it should be 'immutable'). I guess if it needs to be backwards compat for people that are doing that, we can make a setter which changes it nicely, and throws up if it's unicode.

Then we'd have a FilePath._path, which is either bytes or unicode, and the type dictates what "format" it's in, right? And then when you call child, it'll look (roughly) something like this:

```
def child(self, path):
    asPath = self.asBytesPath if type(path) == bytes else self.asTextPath
    return self.clonePath(pathjoin(aspath(), path)) 
```

Windows also uses UTF-16 internally, HOWEVER, standard fopen is *ANSI*. (http://stackoverflow.com/questions/2050973/what-encoding-are-filenames-in-ntfs-stored-as) I am not sure what Python uses.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

https://docs.python.org/2/howto/unicode.html#unicode-filenames leads me to believe that all of Python's stuff (io's open(), os.stat) will accept Unicode filenames and figure out what they should be itself, and do the system-level encoding. Can we get away with just letting Python deal with it?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [hawkowl](#note_5):
> https://docs.python.org/2/howto/unicode.html#unicode-filenames leads me to believe that all of Python's stuff (io's open(), os.stat) will accept Unicode filenames and figure out what they should be itself, and do the system-level encoding. Can we get away with just letting Python deal with it?

Yes and no.

If we just use Unicode paths everywhere, then UNIX users won't be able to open mis-encoded paths any more.  Since `.path` is the _whole_ path, it doesn't even matter if you just do `FilePath(u".").child(u"perfectly-valid-encoding")` you will be unable to call `.open` if any of the current directory's path's parents are not validly encoded.

So we have to hang on to the _ability_ to use bytes paths for starters, and text paths can get progressively smarter later about relative paths and various encoding bugs.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

(In [44075]) Branching to unicode-filepath-7805.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

I did some things. Builders spun.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_9"></a>|@moshez set owner to @hawkowl|
|-|-|

Thanks for working on this. It's a thoughtful API that maintains backwards compat and DTRT under a wide variety of conditions. 

Here are some style improvement suggestions:

"an internal representation of L{bytes}, and vice versa for L{unicode}. " <-- you meant "mutatis mutandis", vice versa should require something that has a direction and goes the other way. For example, if "bytes" meant "asUnicode", vice-versa would be appropriate.

asBytesPath -- you could have it start with "if type(self.path)!=unicode: return self.path" and do all the encoding calculation only when needed? (mutatis mutandis for asTextPath)

In line 734,
couldn't we keep using self.path? You seem to be working a lot to alias self.path to ourPath via confusing means?

In line 1651 -- if slash = slash.encode('ascii') is how we want to spell it, we should do it around line 734 with colon (i.e. "colon = colon.encode('ascii')" )

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

Thanks for the review!

You forgot to remove the review tag to get those juicy highscores points though :)

Will address shortly.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

- I rewrote that section to make it clearer, and expanded upon it.

- Done

- If we use self.path when the internal representation is Unicode and they pass a bytestring, joinpath() will fail ("can't mix bytes and unicode in path components"). We get the goal representation of our internal state (that is, the type of the child path), and then use that, so we only need to encode/decode once.

- I've changed that -- I think it's what you meant. :)


Please review! :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for working on this. Code looks good.

I think that we should push the new public method/members to IFilePath ... but,
at this stage do we need public asBytesPath and asTextPath ?

Since there is no real use case yet for them as public methods maybe we can
start by making them private, but I am not sure that this is a good idea as
we might have soon code needing asBytesPath and asTextPath

-----

I see this code duplicated in current implementation.
Maybe instead of this

```
        if type(pattern) == bytes:
            ourPath = self.asBytesPath()
        else:
            ourPath = self.asTextPath()
```

we can have a helper method

```
ourPath = self._getPathAsSameTypeWith(pattern)
```

------

Do we need to call _getSep(self.path) in sep property each time we read the property? Why not set a private self._sep in __init__ and return that.

Maybe we should also make self.path a read only member to prevent people
from shooting themselves in the foot.

------

I could not find any reference to this in test-standard http://twistedmatrix.com/documents/current/core/development/policy/test-standard.html

but I guess that instead of test_UnicodeInstantiationBytesChild we can have
test_childUnicodeInstantiationBytes to make it clear that we are testing the
child method.

Also, I prefer to format tests using Arrange Act Assert http://c2.com/cgi/wiki?ArrangeActAssert as I find this easier to read.

------

For tests, instead of using ``` u'tmp' ``` maybe we can use a unicode
value which has no equivalent to ASCII ex ```u'mon\u20acy'``` and instead
of ``` b'./tmp' ``` we can use ``` u'parent-mon\\xe2\\x82\\xacy'.encode('utf-8') ```.

With current values I could see test passing on py2.7 OR py3.3 with invalid
implementations of asBytesPath / asTextPath / _getSep and probably for other methods.
Current tests will not pass on py2.7 AND py3.3 but I still that that is nice to stop errors by using a single python version.

I have attached a diff with my updated test which will fail for invalid implementations of asBytesPath and asTextPath while the current tests will pass.

----

As a follow up maybe we can create a ticket to update setContent to accept
Unicode for `ext` argument... and another ticket to accept Unicode for content + encoding.

Thanks for the work on this. Looking forward to see it merge and continue the work on Py3 migration!

Please take a look at my patch for testing and let me know if you find it useful.

needs-changes

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

Thanks Adi!

- I don't think these need to go on IFilePath -- I'm not sure.

- Reduced the code duplication. I added two new (public) functions so that user code could get the benefits of otherwise internal FilePath coercing improvements.

- I think self.path should be too, but that will require deprecation (and is outside of this ticket)

- Thanks for the patch, I applied it. Does much better workouts of it.

- setContent does accept unicode for the ext argument, as it calls to TemporarySibling that returns a unicode FilePath if it's given unicode. And the file is always opened in binary mode (as it should) so things will need explicit encoding to write to it.

- Thanks to idnar on IRC, I figured out why the 10.6 builder was failing. Now it doesn't!

Builders spun, please review!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban commented|
|-|-|

In twisted/python/filepath.py

* what's with all those newlines before _CREATE_FLAGS
* do we really need to export asFilesystemBytes, asFilesystemText asBytesPath asTextPath as public API?
* I don't see any tests for the new code introducing os.path.abspath warning on python 2.6.5. It looks like a different issue to the scope of this ticket. I know that a warning is better but maybe at this stage we can just update the FilePath's docstring to talk about python < 2.6.5 problems.

In twisted/test/test_paths.py

* Instead of ``` u'child-mon\u20acy'.encode('utf-8') ``` maybe we can just have ``` b'child-mon\xe2\x82\xacy' ``` ... I think is easier to read that that we have bytes there.


------

So, I am -1 for updating the public API with 4 new functions/methods at this stage but I think that overall the changes are good so +1 for merging this.

I will leave this in the review queue as I think that the changes are important and there should be at least another dev to look at them.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl commented|
|-|-|

Thanks Adi!

I fixed up some of the newlines, and as mentioned on IRC, I think that the new additions to public API are useful, as it means user code could benefit from further encoding improvements to FilePath.

I also am not sure about the bytes literals -- it makes it more explicit that it's a UTF-8 string. Might come in handy if this explodes on any new platofmrs.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_18"></a>|@hynek set owner to @hawkowl|
|-|-|

twisted/python/filepath.py:

- the docstring of `asFilesystemText` doesn’t make sense to me: `suitable for use on this system's filesystem` makes only sense for bytes, no?
- there code duplication around the handling of sep.  We add a sep property but then don’t use it in `child`, `globChildren` – any particular reason?
 - speaking of which: in child there are two spaces between = and _ (shouldn’t this be caught by pyflakes? hmm)
- I’m -1 on adding `asFilesystemBytes` and `asFilesystemText` to our public API.  Public APIs = maintenance + longer, more confusing API docs.  They are exposed on the FilePath object, that should do.  If it turns out, that they are needed, we can still make them public.  The inverse action is virtually impossible.

optional: the news fragment speaks about text which is of course correct but I feel like we should sneak in the word Unicode somewhere to make clear to outsiders what the hell that means.  Our docstrings also speak about bytes and unicode, not bytes and text (which is pity :))


----

I’ve spun up the builders once more, if they’re green and everyone else happy, feel free to merge after addressing the main points.  As for the last one: you can use glyph or jp as tie breaker I guess.  If you decide to remove them from the API, make sure to re-run the builders.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl commented|
|-|-|

Hi Hynek!

- Hmm, maybe it doesn't. I'll revisit that.
- That's because self.sep is for the filepath's *current* representation -- but if you have a Unicode FP and use a bytes argument, the sep needs to be bytes, while self.sep would return Unicode.
- Okay, I'm convinced. I'll remove the asFilesystem* from the public API.
- I'll revisit the docstring.

Thanks for the review! :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

Just chiming in with my 2¢: I definitely agree with hynek about not expanding the public interface more than necessary.  Hopefully `_asFilesystem(Bytes|Text)` can remain an internal implementation detail forever :).

However, even though it is an internal implementation detail, that parenthetical "probably" is a crack through which some truly abyssal horrors can slither.  It might be good to qualify that "probably" or at least have some central reference in the docstrings themselves to refer to to explain the difficulty of selecting a "real" underlying storage representation.  However, the docstring should at least answer (A) why is it "probably" and (B) why is it in parentheses?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

Oh, also, it would be nice to change this chunk of `FilePath`'s docstring:

```
    The type of C{path} when instantiating decides the internal representation
    of the L{FilePath}. That is, C{FilePath(b"/")} will return a L{bytes} mode
    L{FilePath}, and C{FilePath(u"/")} will return a L{unicode} mode
    L{FilePath}. C{FilePath("/")} will return a L{bytes} mode L{FilePath} on
    Python 2, and a L{unicode} mode L{FilePath} on Python 3.
```

Particularly I object to the phrase "internal representation"; there is an externally-visible consequence to this, whereas the "internal representation" is something that the user might not see.  (And, in the future, I'd like the "internal representation" to be platform-dependent; on Windows and maybe on OS X, the _right_ "internal representation" might be always unicode, with a flag saying "give bytes back to the user").

I think just changing those two words to "mode" might address the problem well enough… but the `@type` annotations throughout the file should change to reference the "mode" of the `FilePath`.

Finally, I had originally intended for `asBytesPath` and `asTextPath` to return a `FilePath` objects, not string types, just in a different mode.  This may be important in the future, where `.asTextPath().basename()` preserves more information and can work more correctly in more scenarios than `FilePath(.asText()).basename()` can.  I think this is actually a pretty important detail.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_22"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44245]) Merging unicode-filepath-7805: proper Unicode support for FilePath

Author: hawkowl
Reviewers: moshez, adiroiban, hynek, glyph
Fixes: [#7805](https://github.com/twisted/twisted/issues/7805)
